flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java

```diff
@@ -146,16 +146,24 @@ void checkAsyncException() {
 		private final Output<StreamRecord<T>> output;
 		private final StreamRecord<T> reuse;
 
-		public NonTimestampContext(StreamSource<?, ?> owner, Object lockingObject, Output<StreamRecord<T>> output) {
-			this.owner = owner;
+		public NonTimestampContext(AbstractStreamOperator<T> owner, Object lockingObject, Output<StreamRecord<T>> output) {
 			this.lockingObject = lockingObject;
 			this.output = output;
 			this.reuse = new StreamRecord<T>(null);
+
+			// if it is a source, then we cast and cache it
+			// here so that we do not have to do it in every collect(),
+			// collectWithTimestamp() and emitWatermark()
+
+			this.owner = (owner instanceof StreamSource) ?
+				(StreamSource) owner : null;
```

Comment:
This looks a bit hacky. How about you add an interface `AsyncException` or so, that all classes the `NonTimestampContext` are using can use.